modules/core/include/opencv2/core/hal/intrin_vsx.hpp

```diff
@@ -1322,11 +1310,105 @@ inline void v_cleanup() {}
 
 ////////// Matrix operations /////////
 
-inline v_int32x4 v_dotprod(const v_int16x8& a, const v_int16x8& b)
-{ return v_int32x4(vec_msum(a.val, b.val, vec_int4_z)); }
+// 16 >> 32
+inline v_int32x4 v_dotprod(const v_int16x8& a, const v_int16x8& b, const bool ignore_order = false)
+{
+    CV_UNUSED(ignore_order);
+    return v_int32x4(vec_msum(a.val, b.val, vec_int4_z));
+}
+inline v_int32x4 v_dotprod(const v_int16x8& a, const v_int16x8& b,
+                           const v_int32x4& c, const bool ignore_order = false)
+{
+    CV_UNUSED(ignore_order);
+    // `vmsumshm` and `vmsumubm` showing a lake of performance when
+    // when the same register is used in `VRT` and `VRC`
+    return v_int32x4(vec_msum(a.val, b.val, vec_int4_z)) + c;
```

Comment:
@pmur, @ChipKerchner , Have anyone faced this issue before? I couldn't find any notes ISA 2.7 or 3 and It seems for me a CPU issue. I tested on both Power8 & 9(little-endian) KVM/VM.

Reply:
I'll ask around. Is using vec_int4_z a workaround to avoid generating those instructions?

Reply:
@pmur, well let me explain it in a better way,
normally we should implement it as following which the third parm of dotprod go directly to vec_msum
```C
return v_int32x4(vec_msum(a.val, b.val, c.val));
```
and by assuming that `a`=`rgv1`, `b`=`rgv2`, c=`rgv3` and returning to `r`=`rgv0` the compiler should produce the following instruction
```ASM
vmsumshm v0,v1,v2,v3
```
now every thing is fine and it shows a great performance but if `c` is overlapped to `r` the following is going to produce
```ASM
vmsumshm v0,v1,v2,v0
```
and depend on many tests I found a missive loss of performance

as a quick fix, I passed `vec_int4_z (splats(0))` to the third element and then add the product to `c` and that should produce three instructions instead of one as following and still shows a better performance

```ASM
vspltisw v4,0
vmsumshm v0, v1, v2, v4
vadduwm v4, v0, v4
```
You can try it by your self, just remove the workaround from all v_dotprod&_expand and compare the peformance u will find more than 30% loss

Reply:
the following benchmark between with and without the workaround shows how can two instructions `vmsumshm` and `vadduwm` since `vspltisw 0` already loaded once, can be faster than one instruction `vmsumshm` when the same register is used in `VRT` and `VRC`

|Name of Test|Without|With|With vs Without (x-factor)|
|---|:-:|:-:|:-:|
|dot::MatType_Length::(8UC1, 32)|0.000|0.000|1.04|
|dot::MatType_Length::(8UC1, 64)|0.001|0.001|1.26|
|dot::MatType_Length::(8UC1, 128)|0.002|0.002|1.28|
|dot::MatType_Length::(8UC1, 256)|0.008|0.006|1.31|
|dot::MatType_Length::(8UC1, 512)|0.033|0.025|1.32|
|dot::MatType_Length::(8UC1, 1024)|0.129|0.098|1.31|
|dot::MatType_Length::(8SC1, 32)|0.000|0.000|1.21|
|dot::MatType_Length::(8SC1, 64)|0.001|0.001|1.33|
|dot::MatType_Length::(8SC1, 128)|0.004|0.003|1.37|
|dot::MatType_Length::(8SC1, 256)|0.015|0.011|1.39|
|dot::MatType_Length::(8SC1, 512)|0.061|0.044|1.39|
|dot::MatType_Length::(8SC1, 1024)|0.243|0.174|1.39

##### "without workaround" patch could be found [here](https://github.com/seiko2plus/opencv/tree/issue15506_revoke_workaround)

Reply:
I agree this workaround improves the dot product operation significantly. However, I don't think the issue is VRT==VRC. This seems to be a second-order effect due to the very small kernel getting generated.

Reply:
well, I did another test for all the instruction set(`vmsummbm`, `vmsumshm`, `vmsumshs`, `vmsumubm`, `vmsumuhm`, `vmsumuhs`) and the situation now changed to worse.
Now we loss performance if the same register is that used in VRT also used in any of (VRA, VRB, VRC).
please, could you check the test [source code](https://raw.githubusercontent.com/seiko2plus/just4share/master/opencv/pr15510/test_msum.cpp) and tell me if I'm missed something.

Reply:
@seiko2plus thanks. I think I see what is going on. For a very small kernel (e.g the 8uC1 above), it's pretty much just a chain of dependent instructions.

What you're seeing is the extra latency of serializing more expensive instructions. Integer vector adds are lower latency, and thus you see a proportional speedup.

This is likely less desirable for a more complicated kernel, but for now I think you can leave the workaround as-is. I would recommend updated the comment however.

Reply:
@pmr, if u checked the test file that I mentioned before you may realize that we really face strange behavior and I still have doubts that it may be a hardware issue and workaround should be from compiler level, however, I removed the comment because I 'm not 100% about sure about the reason, also I made some change on code, so workaround only works when `ignore_order` is true because that may affect negatively on other cases. I will try to investigate this issue deeply when I got free time. thank u

Reply:
I ran a slightly modified version of your example program through the P9 pipeline simulator to characterize the behavior. Moving the data dependency to vadd gives the core freedom to execute the vmsum* sooner, thus hiding it's higher latency in these examples.

If you haven't run across it yet: https://openpowerfoundation.org/?resource_lib=power9-processor-users-manual describes instruction behavior in great detail.

Reply:
> I ran a slightly modified version of your example program through the P9 pipeline simulator to characterize the behavior. Moving the data dependency to vadd gives the core freedom to execute the vmsum* sooner, thus hiding it's higher latency in these examples.

this could be reasonable but wait when I add vadd to vmsum in the case of  `VRT==VRC` still losing latency, I did it in this [test](https://raw.githubusercontent.com/seiko2plus/just4share/master/opencv/pr15510/test_msum.cpp) and here's the result.

Latency of v_msum(vmsummbm) : 186ms
Latency of v_msum_vadd(vmsummbm) : 238ms
Latency of v_msum_no_overlap(vmsummbm) : 53ms
Latency of v_msum_vadd_no_overlap(vmsummbm) : 61ms


> If you haven't run across it yet: https://openpowerfoundation.org/?resource_lib=power9-processor-users-manual describes instruction behavior in great detail.

sure I will look at it hopefully, I can find answer satisfying me.